src/pk-transaction.c

```diff
@@ -1232,6 +1233,143 @@ pk_transaction_package_cb (PkBackend *backend,
+		PkInfoEnum info;
+		PkInfoEnum update_severity;
+		const gchar *package_id;
+		const gchar *summary = NULL;
+		guint encoded_value;
+
+		/* check the backend is doing the right thing */
+		info = pk_package_get_info (item);
+		if (transaction->priv->role == PK_ROLE_ENUM_INSTALL_PACKAGES ||
+		    transaction->priv->role == PK_ROLE_ENUM_UPDATE_PACKAGES) {
+			if (info == PK_INFO_ENUM_INSTALLED) {
+				role_text = pk_role_enum_to_string (transaction->priv->role);
+				g_warning ("%s emitted 'installed' rather than 'installing'",
+					   role_text);
+				continue;
+			}
+		}
+
+		/* check we are respecting the filters */
+		if (pk_bitfield_contain (transaction->priv->cached_filters,
+					 PK_FILTER_ENUM_NOT_INSTALLED)) {
+			if (info == PK_INFO_ENUM_INSTALLED) {
+				role_text = pk_role_enum_to_string (transaction->priv->role);
+				g_warning ("%s emitted package that was installed when "
+					   "the ~installed filter is in place",
+					   role_text);
+				continue;
+			}
+		}
+		if (pk_bitfield_contain (transaction->priv->cached_filters,
+					 PK_FILTER_ENUM_INSTALLED)) {
+			if (info == PK_INFO_ENUM_AVAILABLE) {
+				role_text = pk_role_enum_to_string (transaction->priv->role);
+				g_warning ("%s emitted package that was ~installed when "
+					   "the installed filter is in place",
+					   role_text);
+				continue;
+			}
+		}
+
+		/* add to results even if we already got a result */
+		if (info != PK_INFO_ENUM_FINISHED)
+			pk_results_add_package (transaction->priv->results, item);
+
+		/* emit */
+		package_id = pk_package_get_id (item);
+		g_free (transaction->priv->last_package_id);
+		transaction->priv->last_package_id = g_strdup (package_id);
+		summary = pk_package_get_summary (item);
+		if (transaction->priv->role != PK_ROLE_ENUM_GET_PACKAGES) {
+			g_debug ("emit package %s, %s, %s",
+				 pk_info_enum_to_string (info),
+				 package_id,
+				 summary);
+		}
+
+		/* Safety checks, that the two values do not interleave, neither overflow */
+		g_assert ((PK_INFO_ENUM_LAST & (~0xFFFF)) == 0);
+
+		update_severity = pk_package_get_update_severity (item);
+		encoded_value = info | (((guint32) update_severity) << 16);
+
+		g_variant_builder_add (&builder,
+				       "(uss)",
+				       encoded_value,
+				       package_id,
+				       summary ? summary : "");
+		n_added_packages++;
+	}
+
+	if (n_added_packages == 0) {
+		g_debug ("Empty package array");
+		return;
+	}
+
+	package_array_variant = g_variant_builder_end (&builder);
+
+	/* Emit the signal. Grouping multiple package details into a single
+	 * signal reduces the number of signals and hence the amount of context
+	 * switching between packagekitd, dbus-daemon and the client process.
+	 * This results in much improved performance compared to emitting one
+	 * signal per package.
+	 *
+	 * This should not hit the D-Bus limits (maximum array size of 64MB,
+	 * maximum message size of 128MB) until it’s listing on the order of
+	 * 100000 packages. */
```

Comment:
Should we add a check for that and split the array into multiple `Packages` signals in that case? I got very wary of "should not happen" cases after Fedora started its quest for "how many packages can we build from a texlive distribution?", which made us change default limits in PackageKit multiple times.
Not that I think we will *ever* overflow this during normal operation, but it might be some nice debugging aid if we just fail here with a nice error message.

Reply:
For what it's worth, I believe TeXLive is now at somewhere north of 7k subpackages now, so 10k might actually be too low.

Reply:
> Should we add a check for that and split the array into multiple Packages signals in that case? I got very wary of "should not happen" cases after Fedora started its quest for "how many packages can we build from a texlive distribution?", which made us change default limits in PackageKit multiple times.

I don’t think we should go as far as adding code to split one over-long `Packages` signal into multiple smaller signals. That code would essentially never get tested, so would be full of bugs. Instead, I’ve pushed a fixup commit which falls back to the old loop of `Package` emissions if emitting the `Packages` signal fails (which it will do if the payload is too long).

That means that if we reach 100000 packages, packagekit will regress to emitting single `Package` signals for all of them. But it will still function correctly, and at that point the code can be reworked safely, because it’s being tested.

The fixup commit will need to be squashed in before the PR is merged. I kept it separate for ease of re-review. Aside from squashing it in, this PR is ready to be merged from my PoV.